### DISCUSSION TITLE
How to Keep Only the Longest Notes at Specific Note-On Times

### DISCUSSION THREAD

**Discussion by Jenix-Park** [melanchall/drywetmidi#305] [Q&A] [answered]
Hello, how have you been?

Thanks to DryWetMIDI, I could easily implement a MIDI reader feature.
Over the past few years, I've encountered many MIDI files from different people and realized there are so many MIDI files composed incorrectly.

For example, overlapping notes with zero lengths.
Because their lengths are zero and they are hidden by longer overlapping notes with the same note-on time, they are easily missed by the authors of those MIDI files.

Thankfully, I could easily filter those notes with this one-line code:
`midiFile.RemoveNotes(note => note.Length == 0);`

However, I still encounter MIDI files with overlapping notes whose lengths are greater than zero.
Is there a built-in way to keep only the longest note at a specific note-on time?

For instance, if we have four C4 notes with lengths of 10, 100, 1000, and 1000 **within the same channel** at timestamp 0, I want to remove the 10 and 100 ones and then keep ONLY ONE of the two 1000 notes since they are duplicates.

I want to apply this logic to all notes so that only the longest notes remain **within each channel**.

Is this possible?
Or can you guide me on how to achieve this in the simplest and most efficient way possible?

Thanks!

**Comment by melanchall** [comment, marked as answer]
Hi!

Well, if you don't worry about performance, this approach should work:

```csharp
foreach (var trackChunk in midiFile.GetTrackChunks())
{
    using (var notesManager = trackChunk.ManageNotes())
    {
        var notesGroups = notesManager
            .Objects
            .GroupBy(n => (n.Time, n.NoteNumber, n.Channel));

        foreach (var notesGroup in notesGroups)
        {
            var maxLength = notesGroup.Max(n => n.Length);
            var singleMaxLengthNote = notesGroup.First(n => n.Length == maxLength);

            foreach (var note in notesGroup)
            {
                if (note != singleMaxLengthNote)
                    notesManager.Objects.Remove(note);
            }
        }
    }
}
```

But be aware this code will work only if overlapping notes start at the same time. Because if they don't, things will be non-deterministic. You can read more [here](https://melanchall.github.io/drywetmidi/articles/high-level-managing/Getting-objects.html#notestartdetectionpolicy).

Also I probably should think about adding such functionality in the [Sanitizer](https://melanchall.github.io/drywetmidi/articles/tools/Sanitizer.html) tool. Thank you for interesting case.

**Reply by Jenix-Park** [reply]
Thank you so much for your quick response and the neat solution!

After running some simple tests and reviewing the code, it seems like it won't affect normal MIDI files with no overlapping notes, which is great because it ensures that previous behaviors won't be disrupted.

However, just to be sure, could you confirm that your explanation—"But be aware this code will work only if overlapping notes start at the same time. Because if they don't, things will be non-deterministic."—refers specifically to overlapping notes with the same note and channel numbers but with slight time offsets?

If that's the case, I don't think I need to worry.
In my experience, such overlapping notes aren't usually intentional since they often produce weird and inconsistent results across different MIDI modules and software.
There might be some rare instances where someone uses this as a special trick for doubling effects, but these would likely only work on specific MIDI devices.
Also, from what I've seen in the millions of MIDI files from different people, this has never been the case yet.

So, if the code works for overlapping notes, that's fantastic!
But even if it doesn't handle all cases perfectly, I’m not too concerned since those notes would sound strange anyway.
I'm just grateful that the code helps reduce some artifacts.

Thanks again!

**Reply by Jenix-Park** [reply]
@melanchall 
Oh, by the way, would using the Sanitizer tool be more efficient than my current approach of using `midiFile.RemoveNotes(note => note.Length == 0);`?
I wasn't aware that the Sanitizer could handle what I was trying to do.

**Reply by melanchall** [reply]
@Jenix-Park

> However, just to be sure, could you confirm that your explanation—"But be aware this code will work only if overlapping notes start at the same time. Because if they don't, things will be non-deterministic."—refers specifically to overlapping notes with the same note and channel numbers but with slight time offsets?

Yes, exactly.

> Oh, by the way, would using the Sanitizer tool be more efficient than my current approach of using `midiFile.RemoveNotes(note => note.Length == 0);`?

Hmm, I think no. `Sanitizer` is a great tool to perform multiple cleaning routines simultaneously. If removing zero-length notes is the only thing you want to do, I suppose both approaches have the same performance.

**Reply by Jenix-Park** [reply]
@melanchall 
Oh, thanks for the confirmation!

By the way, I'm currently manually removing notes with a velocity of 0.
I thought the Sanitizer might be able to handle this, but I couldn’t find a setting for it.
Is there an option I might be missing?

**Reply by Jenix-Park** [reply]
I meant a velocity of 1.
Removing notes with a velocity of 0 might cause issues because those events actually represent Note-Off messages, right?

**Reply by melanchall** [reply]
> I thought the Sanitizer might be able to handle this, but I couldn’t find a setting for it.

No, there is no such an option. But I'll add it, thank you! Something like `NoteMinVelocity`. All notes with velocity below this value will be deleted.

**Reply by Jenix-Park** [reply]
Thanks a lot!
DryWetMIDI just keeps improving—I’m loving it more and more! :)

**Reply by melanchall** [reply]
> Removing notes with a velocity of 0 might cause issues because those events actually represent Note-Off messages, right?

Not quite. **Note On events** with zero velocity can be treated as Note Off ones (btw you can manage this behavior with DWM). But **a note** is a pair of Note On and Note Off events. So **a note** with zero velocity is probably junk indeed in many cases.

**Reply by Jenix-Park** [reply]
Wow thanks!😃

I assumed DryWetMIDI would automatically treat notes with zero velocity as Note Off events, so I didn’t think I needed to handle them manually. If that’s not the case, could you point me to where I can configure this behavior?

**Reply by melanchall** [reply]
Yes, if we're talking about reading a MIDI file, DWM treats silent Note On events as Note Off ones. You can control this behavior via [`ReadingSettings.SilentNoteOnPolicy`](https://melanchall.github.io/drywetmidi/api/Melanchall.DryWetMidi.Core.ReadingSettings.html#Melanchall_DryWetMidi_Core_ReadingSettings_SilentNoteOnPolicy) property of the settings passed to `MidiFile.Read` methods.

**Reply by Jenix-Park** [reply]
Thank you for the explanation!
